update: tutorial on publishing a KMP library (#288)
JetBrains/kotlin-multiplatform-dev-docs: topics/multiplatform-publish-libraries.md

brew install gpg
```

1. Start generating a key pair with the following command and fill in the required details when prompted:

    ```bash
    gpg --full-generate-key
    ```

2. Choose the recommended defaults for the type of key to be created.
    You can leave the selections empty and press Enter to accept the default values.

    ```text
    Please select what kind of key you want:
        (1) RSA and RSA
        (2) DSA and Elgamal
        (3) DSA (sign only)
        (4) RSA (sign only)
        (9) ECC (sign and encrypt) *default*
        (10) ECC (sign only)
        (14) Existing key from card
    Your selection? 9
    
    Please select which elliptic curve you want:
        (1) Curve 25519 *default*
        (4) NIST P-384
        (6) Brainpool P-256
    Your selection? 1
    ```

    > At the time of writing, this is `ECC (sign and encrypt)` with `Curve 25519`.
    > Older versions of `gpg` might default to `RSA` with a `3072` bit key size.
    > 
    {style="note"}

3. When prompted to how long the key should be valid, you can choose the default option of no expiration date.
    If you choose to create a key that automatically expires after a set amount of time,

>     you’ll need to [extend its validity](https://central.sonatype.org/publish/requirements/gpg/#dealing-with-expired-keys) when it expires.

Author: I'd say "when it expires" implies that this is going to happen later, no?

// <module directory>/build.gradle.kts

mavenPublishing {
    publishToMavenCentral(SonatypeHost.CENTRAL_PORTAL)
    
    signAllPublications()
    
    coordinates(group.toString(), "fibonacci", version.toString())
    
    pom { 
        name = "Fibonacci library"
        description = "A mathematics calculation library."
        inceptionYear = "2024"
        url = "https://github.com/kotlin-hands-on/fibonacci/"
        licenses {
            license {
                name = "The Apache License, Version 2.0"
                url = "https://www.apache.org/licenses/LICENSE-2.0.txt"
                distribution = "https://www.apache.org/licenses/LICENSE-2.0.txt"
            }
        }
        developers {
            developer {
                id = "kotlin-hands-on"
                name = "Kotlin Developer Advocate"
                url = "https://github.com/kotlin-hands-on/"
            }
        }
        scm {
            url = "https://github.com/kotlin-hands-on/fibonacci/"
            connection = "scm:git:git://github.com/kotlin-hands-on/fibonacci.git"
            developerConnection = "scm:git:ssh://git@github.com/kotlin-hands-on/fibonacci.git"
        }
    }
}
```

> > It’s also possible to use Gradle properties to configure this instead.

Author: Not sure what you mean here, could you suggest some reasoning here? I'm adding the link to the Gradle properties doc because I thought it might not be obvious what that is, but otherwise not sure what to change here.

> 
{style="tip"}

The most important settings here are:

* The `coordinates`, which specify the `groupId`, `artifactId`, and `version` of your library.

> * The [license](https://central.sonatype.org/publish/requirements/#license-information) that you’re publishing your library under.

Reviewer: ```suggestion
* The [license](https://central.sonatype.org/publish/requirements/#license-information) under which your library is published.
```

* The [developer information](https://central.sonatype.org/publish/requirements/#developer-information) which lists the authors of the library.
* [SCM (Source Code Management) information](https://central.sonatype.org/publish/requirements/#scm-information),

>   which specifies where the sources of your library are available.

Reviewer: ```suggestion
  which specifies where the library’s source code is hosted.
```

### Publish to Maven Central using continuous integration

### Generate the user token

You need a Maven access token for Maven Central to authorize your publishing requests.
Open the [Setup Token-Based Authentication](https://central.sonatype.com/account) page and click the **Generate User Token** button.

The output looks like the example below, containing a username and a password.

> If you lose these credentials, you’ll need to generate new ones later — they are not stored by Maven Central.

Reviewer: ```suggestion
If you lose these credentials, you’ll need to generate new ones later, as they are not stored by Maven Central.
```

```xml
<server>
    <id>${server}</id>
    <username>l2nfaPmz</username>
    <password>gh9jT9XfnGtUngWTZwTu/8141keYdmQpipqLPRKeDLTh</password>
</server>
```

### Add secrets to GitHub

To use the keys and credentials required for publication in your GitHub Action workflow while keeping them private,

> you need to place those values into secrets.

Reviewer: ```suggestion
you need to store these values as secrets.
```

1. On your GitHub repository **Settings** page, click **Security** | **Secrets and variables** | **Actions**.
2. Click the `New repository secret` button and add the following secrets:

>     * `MAVEN_CENTRAL_USERNAME` and `MAVEN_CENTRAL_PASSWORD` are the values generated as the User Token by the Central Portal website in the [Generate User Token](#generate-the-user-token) section.

Reviewer: ```suggestion
    * `MAVEN_CENTRAL_USERNAME` and `MAVEN_CENTRAL_PASSWORD` are the values generated for the User Token by the Central Portal website, as described in the [Generate User Token](#generate-the-user-token) section.
```

    * `SIGNING_KEY_ID` is **the last 8 characters** of your signing key’s identifier, for example, `20B385CE` for `F175482952A225BFC4A07A715EE6B5F76620B385CE`

>     * `SIGNING_PASSWORD` is the passphrase you’ve provided when generating your GPG key.

Reviewer: ```suggestion
    * `SIGNING_PASSWORD` is the passphrase you provided when generating your GPG key.
```

>     * `GPG_KEY_CONTENTS` should contain the entire contents of your `key.gpg` file, which you’ve created earlier

Reviewer: ```suggestion
    * `GPG_KEY_CONTENTS` should contain the entire contents of your `key.gpg` file, which you created earlier
```

        in the [Export your private key](#export-your-private-key) section.

![](../images/publish/github_secrets.png)

You will use the names for these secrets in the CI configuration on the next step.

#### Add a GitHub Actions workflow to your project

> You can set up continuous integration which builds and publishes your library automatically.

Reviewer: ```suggestion
You can set up continuous integration to build and publish your library automatically.
```

We’ll use [GitHub Actions](https://docs.github.com/en/actions) as an example.

To get started, add the following workflow to your repository, in the `.github/workflows/publish.yml` file.

```yaml
# .github/workflows/publish.yml

name: Publish
on:
  release:
    types: [released, prereleased]
jobs:
  publish:
    name: Release build and publish
    runs-on: macOS-latest
    steps:
      - name: Check out code
        uses: actions/checkout@v4
      - name: Set up JDK 21
        uses: actions/setup-java@v4
        with:
          distribution: 'zulu'
          java-version: 21
      - name: Publish to MavenCentral
        run: ./gradlew publishToMavenCentral --no-configuration-cache
        env:
          ORG_GRADLE_PROJECT_mavenCentralUsername: ${{ secrets.MAVEN_CENTRAL_USERNAME }}
          ORG_GRADLE_PROJECT_mavenCentralPassword: ${{ secrets.MAVEN_CENTRAL_PASSWORD }}
          ORG_GRADLE_PROJECT_signingInMemoryKeyId: ${{ secrets.SIGNING_KEY_ID }}
          ORG_GRADLE_PROJECT_signingInMemoryKeyPassword: ${{ secrets.SIGNING_PASSWORD }}
          ORG_GRADLE_PROJECT_signingInMemoryKey: ${{ secrets.GPG_KEY_CONTENTS }}
```

> After committing and pushing this file, the workflow will run automatically when you create a release (including a pre-release)

Reviewer: ```suggestion
Once you commit and push this file, the workflow will run automatically whenever you create a release (including a pre-release)
```

in the GitHub repository hosting your project.

> It checks out the current version of your code, sets up a JDK, and then runs the `publishToMavenCentral` Gradle task.

Reviewer: ```suggestion
The workflow checks out the current version of your code, sets up a JDK, and then runs the `publishToMavenCentral` Gradle task.
```

When using the `publishToMavenCentral` task, you’ll still need to check and release your deployment manually on the Maven Central website,
as described in the [](#create-a-release-on-github) section.

> You may use `publishAndReleaseToMavenCentral` instead for a fully automated release.

Reviewer: ```suggestion
Alternatively, you can use the `publishAndReleaseToMavenCentral` task to enable a fully automated release process.
```

You can also configure the workflow to [trigger when a tag is pushed](https://stackoverflow.com/a/61892639) to your repository.

> The script above disables Gradle [configuration cache](https://docs.gradle.org/current/userguide/configuration_cache.html)
> for the publication task by adding `--no-configuration-cache` to the Gradle command,
> as the publication plugin does not support it (see this [open issue](https://github.com/gradle/gradle/issues/22779)).
>
{style="tip"}

> This action will need your signing details and your Maven Central credentials, which you created as repository secrets

Reviewer: ```suggestion
This action requires your signing details and your Maven Central credentials, which you created as repository secrets
```

in the [](#add-secrets-to-github) section.

> The configuration of the workflow transfers the secrets into environment variables,
> which will make them available to the Gradle build automatically.

Reviewer: ```suggestion
The workflow configuration automatically transfers these secrets into environment variables,
making them available to the Gradle build process.
```

#### Create a release on GitHub

With the workflow and secrets set up, you’re now ready to [create a release](https://docs.github.com/en/repositories/releasing-projects-on-github/managing-releases-in-a-repository#creating-a-release)
that will trigger the publication of your library.

1. Make sure that the version number specified in the `build.gradle.kts` file for your library is the one you would like to publish.
2. Go to your GitHub repository’s main page.

> 3. In the right sidebar click **Releases**.

Reviewer: ```suggestion
3. In the right sidebar, click **Releases**.
```

4. Click the **Draft a new release** button (or the **Create a new release** button if you haven't created a release for this repository before).

> 5. Each release creates a new tag. Create a new tag in the tags dropdown, and set the release title (these may be identical).

Reviewer: ```suggestion
5. Each release creates a new tag. Create a new tag in the tags dropdown, and set the release title (these values are often identical).
```

Author: The intended message is "you can make them the same", I'll clarify that.

   You probably want these to be the same as the version number of the library that you specified in the `build.gradle.kts` file.

    ![](../images/publish/create_release_and_tag.png)

> 6. Double-check the branch you want to target with the release (especially if you want to release from a branch that’s different from your default),

Reviewer: ```suggestion
6. Double-check the branch you want to target with the release (especially if it’s not the default branch),
```

    and add appropriate release notes for your new version.

> 7. The checkboxes below the description allow you to mark a release as a pre-release (useful for early access – alpha, beta, or RC versions of a library).

Reviewer: ```suggestion
7. Use the checkboxes below the description to mark a release as a pre-release (useful for early access versions such as alpha, beta, or RC).
```

   You can also mark the release as the latest one (if you already made a release for this repository before):

8. Click the **Publish release** button to create the new release.

9. Click the **Actions** tab at the top of your GitHub repository's page.
    Here you’ll see that the new release triggered your publishing workflow.
    Click on it to see the outputs of the publication task.

10. After the workflow run completes successfully, navigate to the [Deployments](https://central.sonatype.com/publishing/deployments)
     dashboard on Maven Central.

>      You should see a new deployment here. This deployment can be in the *pending* or *validating* states for some time

Reviewer: ```suggestion
     You should see a new deployment here. This deployment may remain in the *pending* or *validating* states for some time
```

     while Maven Central performs checks on it.

> Once your deployment is in the *validated* state, you should check that it contains all the artifacts you’ve uploaded.

Reviewer: ```suggestion
Once your deployment is in the *validated* state, verify that it contains all the artifacts you uploaded.
```

> If everything looks correct, click the *Publish* button to release these artifacts.

Reviewer: ```suggestion
If everything is correct, click the *Publish* button to release these artifacts.
```

![](../images/publish/published_on_maven_central.png)

> > It will take some time (usually about 15–30 minutes) after the release for the artifacts to be available publicly in the Maven Central repository.

Reviewer: ```suggestion
> It will take some time (usually about 15–30 minutes) after the release for the artifacts to become publicly available in the Maven Central repository.
```

> > It may take some more time before they are indexed to be found on [the Maven Central website](https://central.sonatype.com/).

Reviewer: ```suggestion
> Additionally, it may take longer for them to be indexed and become searchable on [the Maven Central website](https://central.sonatype.com/).
```

> To release the artifacts automatically once the deployment is verified, replace the `publishToMavenCentral` task in your workflow

Reviewer: ```suggestion
To automate the release of artifacts once the deployment is verified, replace the `publishToMavenCentral` task in your workflow
```
